modules/packages/Python.chpl

```diff
@@ -0,0 +1,1417 @@
+
+      .. note::
+
+         Most users should not need to call this directly, except when writing
+         a :type:`TypeConverter`.
+    */
+    proc fromPython(type t, obj: c_ptr(void)): t throws {
+      for converter in this.converters {
+        if converter.handlesType(t) {
+          return converter.fromPython(this, t, obj);
+        }
+      }
+      // if no converter is found, use the defaults
+      if isIntType(t) {
+        var v = PyLong_AsLong(obj);
+        this.checkException();
+        return v: t; // cast to the real type
+      } if isUintType(t) {
+        var v = PyLong_AsUnsignedLong(obj);
+        this.checkException();
+        return v: t; // cast to the real type
+      } else if isRealType(t) {
+        var v = PyFloat_AsDouble(obj);
+        this.checkException();
+        return v: t; // cast to the real type
+      } else if isBoolType(t) {
+        return if obj == Py_True then true else false;
+      } else if t == c_ptrConst(c_char) {
+        var v = PyUnicode_AsUTF8(obj);
+        this.checkException();
+        return v;
+      } else if t == string {
+        var v = string.createCopyingBuffer(PyUnicode_AsUTF8(obj));
+        this.checkException();
+        return v;
+      } else if isArrayType(t) {
+        return fromList(t, obj);
+      } else if isSubtype(t, List.list) {
+        return fromList(t, obj);
+      } else if isSubtype(t, Function) {
+        return new Function("<unknown>", new PyObject(this, obj));
+      } else if isSubtype(t, PyObject) || isSubtype(t, PyObject?) {
+        return new PyObject(this, obj);
+      } else if isSubtype(t, ClassObject) || isSubtype(t, ClassObject?) {
+        return new ClassObject(new PyObject(this, obj));
+      } else if t == NoneType {
+        return None;
+      } else {
+        halt("Unsupported fromPython type: '" + t:string + "'");
+      }
+    }
+
+    /*
+      Converts an array to a python list
+    */
+    @chpldoc.nodoc
+    proc toList(arr): c_ptr(void) throws
+      where isArrayType(arr.type) && arr.rank == 1 && arr.isDefaultRectangular() {
+      var pyList = PyList_New(arr.size);
+      this.checkException();
+      for i in 0..<arr.size {
+        const idx = arr.domain.orderToIndex(i);
+        PyList_SetItem(pyList, i, toPython(arr[idx]));
+        this.checkException();
+      }
+      return pyList;
+    }
+
+    /*
+      Convert a chapel list to a python list
+    */
+    @chpldoc.nodoc
+    proc toList(l: List.list(?)): c_ptr(void) throws {
+      var pyList = PyList_New(l.size);
+      this.checkException();
+      for i in 0..<l.size {
+        PyList_SetItem(pyList, i, toPython(l(i)));
+        this.checkException();
+      }
+      return pyList;
+    }
+
+    /*
+      Converts a python list to an array
+    */
+    @chpldoc.nodoc
+    proc fromList(type T, obj: c_ptr(void)): T throws
+      where isArrayType(T) {
+
+      if (PySequence_Check(obj)) {
```

Comment:
Consider flattening this as:

```Chapel
if !PySequence_Check(obj) {
  throw ...
}

// the rest...
```

The large bodies of the if/else make it hard to follow.